### PR TITLE
feature(delete button): add button to delete table row

### LINKED ISSUE #19: Add ability to delete a record

### PR DESCRIPTION
**Pull request by ThomastheBicyclist**
closes #19

### CODE REVIEW

File: src/views/record-browser/table-data/table-data.html [hunk 1]
```diff
@@ -1,15 +1,23 @@
 <table>
 	<thead>
 	<tr>
+		<th></th>
```
Comment (nick-herrera) [inline src/views/record-browser/table-data/table-data.html:4]: This table heading needs `class="actions"` to match the table data it describes. If a table has no records, the width of this column is undefined.

image

File: src/api/document.service.js [hunk 2]
```diff
@@ -5,13 +5,20 @@
 		.module('montage')
 		.factory('documentService', documentService);
 
-	function documentService($http, authService) {
+	function documentService($http, authService, montageHelper) {
+		// console.log(montageHelper.getClient());
```
Comment (nick-herrera) [inline src/api/document.service.js:9]: Remove unused code

File: src/services/toast.service.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+(function (angular) {
+	'use strict';
+
+	angular
+		.module('montage')
+		.factory('toast', toastService);
+
+	function toastService($mdToast) {
+		return {
+	  	simple
```
Comment (nick-herrera) [inline src/services/toast.service.js:10]: Incorrect indentation level. Also, there is a mixture of spaces and tabs here. Stick to tabs.

File: src/services/toast.service.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+		}
```
Comment (nick-herrera) [inline src/services/toast.service.js:11]: style: missing semi-colon (yes, they are optional, but we use them)

File: src/services/toast.service.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+		function simple(message){
+			$mdToast.show($mdToast.simple()
+			.textContent(message)
+			.action("Dismiss"));
```
Comment (nick-herrera) [inline src/services/toast.service.js:16]: Let's use "OK" instead of "Dismiss"

File: src/services/toast.service.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+		.module('montage')
+		.factory('toast', toastService);
+
+	function toastService($mdToast) {
+		return {
+	  	simple
+		}
+
+		function simple(message){
+			$mdToast.show($mdToast.simple()
```
Comment (nick-herrera) [inline src/services/toast.service.js:14]: I feel like this is more readable as one of the following.

``` javascript
$mdToast.show(
    $mdToast.simple()
    .textContent(message)
    .action("OK")
);
```

``` javascript
$mdToast.show($mdToast.simple()
    .textContent(message)
    .action("OK")
);
```

File: src/views/record-browser/query-builder/query-builder.component.js [hunk 7]
```diff
@@ -44,6 +44,8 @@
 			};
 		};
 
+
```
Comment (nick-herrera) [inline src/views/record-browser/query-builder/query-builder.component.js:47]: style: unnecessary line additions

File: src/views/record-browser/query-builder/query-builder.component.js [hunk 8]
```diff
@@ -86,6 +88,8 @@
 
 			var query = new montage.Query(schema);
 
+			$scope.schemaName = schema;
```
Comment (nick-herrera) [inline src/views/record-browser/query-builder/query-builder.component.js:91]: `$scope.schemaName` is not used. Remove these two lines.

File: src/views/record-browser/query-builder/query-builder.html [hunk 9]
```diff
@@ -67,13 +67,13 @@
 
 	<label>Filters</label>
 
-	<button
+	<div
```
Comment (nick-herrera) [inline src/views/record-browser/query-builder/query-builder.html:70]: This fix is unrelated to the issue that this feature request (#19) addresses. In the future, create a new issue and a separate pull-request.

File: src/views/record-browser/query-builder/query-builder.scss [hunk 10]
```diff
@@ -28,6 +28,7 @@
 		border-radius: 50%;
 		color: white;
 		float: right;
+		cursor: pointer;
```
Comment (nick-herrera) [inline src/views/record-browser/query-builder/query-builder.scss:31]: Unrelated fix (continued)

File: src/views/record-browser/record-browser.component.js [hunk 11]
```diff
@@ -26,6 +26,7 @@
 		};
 
 		function getSchema(schemaName) {
+			// $scope.schemaName = vm.schemaList.filter(schema => schema.name === schemaName)[0];
```
Comment (nick-herrera) [inline src/views/record-browser/record-browser.component.js:29]: Remove unused code

File: src/views/record-browser/record-browser.component.js [hunk 12]
```diff
@@ -9,7 +9,7 @@
 			controller: recordBrowserController
 		});
 
-	function recordBrowserController(api, montageHelper) {
+	function recordBrowserController(api, montageHelper, $scope, $window) {
```
Comment (nick-herrera) [inline src/views/record-browser/record-browser.component.js:12]: Remove unused dependencies.

File: src/views/record-browser/record-browser.html [hunk 13]
```diff
@@ -8,12 +8,12 @@
 	<div class="help-text" ng-hide="recordBrowser.results">
 		<h1>Use the Query Builder to the left to retrieve your data.</h1>
 	</div>
-
```
Comment (nick-herrera) [inline src/views/record-browser/record-browser.html:11]: Unnecessary change. If this is intentional, it should be its own commit.

For example:
`style(record-browser): remove unnecessary blank line`

File: src/views/record-browser/record-browser.scss [hunk 14]
```diff
@@ -62,4 +62,57 @@ $sidebar-width: 300px;
 			&.sep { color: $gray-medium; }
 		}
 	}
+
+	thead {
+		/* remove formatting for delete button column header cell */
+		th:first-child {
+			color: transparent;
+			background: transparent;
+			border-bottom: 0px;
```
Comment (nick-herrera) [inline src/views/record-browser/record-browser.scss:71]: Units are unnecessary when the value is zero. `0`, `0em`, `0px`, `0%`, etc are all equivalent

File: src/views/record-browser/table-data/table-data.component.js [hunk 15]
```diff
@@ -9,11 +9,12 @@
...
 			bindings: {
 				schemaFields: '=',
-				data: '='
+				data: '=',
+				schemaName: '='
 			}
 		});
 
-	function tableDataController($scope) {
+	function tableDataController($scope, $mdDialog, modalHelper, api, toast) {
```
Comment (nick-herrera) [inline src/views/record-browser/table-data/table-data.component.js:17]: `$mdDialog` is not used. Remove it from the list of dependencies.

File: src/views/record-browser/table-data/table-data.component.js [hunk 16]
```diff
@@ -67,9 +68,35 @@
...
+						}
+						// see src/services/toast.service.js
+						toast.simple("Record deleted successfully");
+					})
+					.catch(function (error) {
+						console.error("Error during delete: " + error);
+						toast.simple("There was an error deleting the record.");
+					});
+			}
+			, function() {
```
Comment (nick-herrera) [inline src/views/record-browser/table-data/table-data.component.js:96]: style: I've never seen a comma lead a line before. But then again, I don't use this style of error handling. I would have written this as: `}, function() {`

Also, above you use the `.then().catch()` format, but here you use the `.then(success, error)` format. Strive for consistency

File: src/views/record-browser/table-data/table-data.component.js [hunk 17]
```diff
@@ -67,9 +68,35 @@
...
-
 			// Convert document objects into arrays with the same order as `columns`
 			return data.map(document => columns.map(column => document[column.name]));
 		}
+
+		// Modal that pops up when clicking on delete row icon
+		$scope.showConfirm = function(row) {
+			var result = modalHelper.confirmDelete('record');
+			var row_id = row[0];
+			result.then(function() {
```
Comment (nick-herrera) [inline src/views/record-browser/table-data/table-data.component.js:79]: `result` could be named something more descriptive.  However, it makes sense to get rid of the variable altogether and write `modalHelper.confirmDelete('record').then(...)`

File: src/views/record-browser/table-data/table-data.component.js [hunk 18]
```diff
@@ -67,9 +68,35 @@
...
+				// deletes row from database but not the view
+				api.document.remove(vm.schemaName, row_id)
+				// deletes row from view but not the database
+					.then(function (response) {
```
Comment (nick-herrera) [inline src/views/record-browser/table-data/table-data.component.js:83]: remove unused variable `response`

File: src/views/record-browser/table-data/table-data.component.js [hunk 19]
```diff
@@ -67,9 +68,35 @@
...
+						var row_idIndex = vm.rows.indexOf(row);
+						if (row_idIndex > -1) {
+							vm.rows.splice(row_idIndex, 1);
+						}
+						// see src/services/toast.service.js
```
Comment (nick-herrera) [inline src/views/record-browser/table-data/table-data.component.js:88]: This is implied. Furthermore, the path may change and make this comment invalid.

File: src/views/record-browser/table-data/table-data.component.js [hunk 20]
```diff
@@ -67,9 +68,35 @@
...
+						toast.simple("Record deleted successfully");
+					})
+					.catch(function (error) {
+						console.error("Error during delete: " + error);
+						toast.simple("There was an error deleting the record.");
+					});
+			}
+			, function() {
+				// If user clicks cancel button this happens
+				toast.simple("Delete record cancelled.");
```
Comment (nick-herrera) [inline src/views/record-browser/table-data/table-data.component.js:98]: "Delete canceled" sounds better. However, is this necessary at all? The user knows they clicked cancel.

Side note: `canceled` is the American spelling of `cancelled`. (I prefer the double `l` too, though)

File: src/views/record-browser/table-data/table-data.html [hunk 21]
```diff
@@ -1,15 +1,23 @@
...
 		<th
 			ng-repeat="column in tableData.columns"
 			ng-class="{ secondary : column.nonSchema }"
 		>{{ ::column.name }}</th>
 	</tr>
 	</thead>
 	<tbody>
-	<tr ng-repeat="row in tableData.rows">
-		<td ng-if="!tableData.isJson(cell)" ng-repeat="cell in ::row track by $index">{{ ::cell }}</td>
+	<tr ng-repeat="row in tableData.rows" ng-mouseenter="showDeleteButton = true" ng-mouseleave="showDeleteButton = false">
```
Comment (nick-herrera) [inline src/views/record-browser/table-data/table-data.html:12]: Use CSS to show/hide the delete button

File: src/views/record-browser/record-browser.scss [hunk 22]
```diff
@@ -62,4 +62,57 @@ $sidebar-width: 300px;
 			&.sep { color: $gray-medium; }
 		}
 	}
+
+	thead {
```
Comment (nick-herrera) [inline src/views/record-browser/record-browser.scss:66]: All of these table styles are specific to the `table-data` component. Move them to `src/views/record-browser/table-data/table-data.scss`

File: src/views/record-browser/record-browser.scss [hunk 23]
```diff
@@ -62,4 +62,57 @@ $sidebar-width: 300px;
...
+			border-bottom: 0px;
+			border-left: 0px;
+		}
+		/* aligns column seperator line for the 2nd column header cell */
+		th:nth-child(2) {
+			border-left: $blue-gray-light;
+		}
+	}
+
+	/* remove border from row, so it can be applied per cell */
```
Comment (nick-herrera) [inline src/views/record-browser/record-browser.scss:80]: Consider making this change in `src/static/styles/tables.scss`

File: src/views/record-browser/record-browser.scss [hunk 24]
```diff
@@ -62,4 +62,57 @@ $sidebar-width: 300px;
 			&.sep { color: $gray-medium; }
 		}
 	}
+
+	thead {
+		/* remove formatting for delete button column header cell */
+		th:first-child {
```
Comment (nick-herrera) [inline src/views/record-browser/record-browser.scss:68]: At the component level, we prefer using class selectors instead of tag names.

File: src/views/record-browser/record-browser.scss [hunk 25]
```diff
@@ -62,4 +62,57 @@ $sidebar-width: 300px;
...
+			color: transparent;
+			background: transparent;
+			border-bottom: 0px;
+			border-left: 0px;
+		}
+		/* aligns column seperator line for the 2nd column header cell */
```
Comment (nick-herrera) [inline src/views/record-browser/record-browser.scss:74]: `seperator` is spelled `separator`

File: src/views/record-browser/record-browser.scss [hunk 26]
```diff
@@ -62,4 +62,57 @@ $sidebar-width: 300px;
...
+	/* format delete button icon */
+	.fa {
+		color: $gray-mid !important;
+	}
+	.actions:hover .fa {
+		color: $gray-dark !important;
+	}
+	.actions:active .fa {
+		color: $blue !important;
+		border: transparent;
```
Comment (nick-herrera) [inline src/views/record-browser/record-browser.scss:111]: This line doesn't do anything. Were you trying to do the following?

``` css
.delete-record-button:active {
  outline: 0;
}
```

**Comment by nick-herrera** [comment]
In the future, please make an effort to make atomic commits. There was only a single commit for this pull request containing 11 changed files and several distinct tasks.

Overall, it seems that the issue's requirements were met.

There are a few things I'd like you to fix and several less important stylistic issues that won't prevent this pull request from being merged in.

The important bits are:
- [x] Fix the styling issue present when there are no records.
- [x] Move the table styles in to the `table-data` directory
- [x] Remove the `outline` on `.delete-record-button:active`
- [x] Remove commented-out code and unused variable references.

After these items are addressed, I will probably review your CSS one more time before merging.

File: src/views/record-browser/table-data/table-data.html (hunk 1 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/table-data/table-data.html:4]: Done.

File: src/api/document.service.js (hunk 2 above)
Comment (ThomastheBicyclist) [inline src/api/document.service.js:9]: Done.

File: src/services/toast.service.js (hunk 3 above)
Comment (ThomastheBicyclist) [inline src/services/toast.service.js:10]: Done.

File: src/services/toast.service.js (hunk 5 above)
Comment (ThomastheBicyclist) [inline src/services/toast.service.js:16]: Done.

File: src/services/toast.service.js (hunk 6 above)
Comment (ThomastheBicyclist) [inline src/services/toast.service.js:14]: Done.

File: src/views/record-browser/query-builder/query-builder.component.js (hunk 7 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/query-builder/query-builder.component.js:47]: Done.

File: src/views/record-browser/query-builder/query-builder.component.js (hunk 8 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/query-builder/query-builder.component.js:91]: Done.

File: src/views/record-browser/query-builder/query-builder.html (hunk 9 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/query-builder/query-builder.html:70]: Will do.

File: src/views/record-browser/record-browser.component.js (hunk 11 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/record-browser.component.js:29]: Done.

File: src/views/record-browser/record-browser.component.js (hunk 12 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/record-browser.component.js:12]: Done.

File: src/views/record-browser/record-browser.html (hunk 13 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/record-browser.html:11]: Done.

File: src/views/record-browser/record-browser.scss (hunk 22 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/record-browser.scss:66]: Done.

File: src/views/record-browser/record-browser.scss (hunk 24 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/record-browser.scss:68]: Done.

File: src/views/record-browser/record-browser.scss (hunk 14 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/record-browser.scss:71]: Done.

File: src/views/record-browser/record-browser.scss (hunk 26 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/record-browser.scss:111]: Removed.

File: src/views/record-browser/table-data/table-data.component.js (hunk 15 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/table-data/table-data.component.js:17]: Done.

File: src/views/record-browser/table-data/table-data.component.js (hunk 17 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/table-data/table-data.component.js:79]: Done.

File: src/views/record-browser/table-data/table-data.component.js (hunk 18 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/table-data/table-data.component.js:83]: Done.

File: src/views/record-browser/table-data/table-data.component.js (hunk 19 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/table-data/table-data.component.js:88]: Done.

File: src/views/record-browser/table-data/table-data.component.js (hunk 16 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/table-data/table-data.component.js:96]: Done.

File: src/views/record-browser/table-data/table-data.component.js (hunk 20 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/table-data/table-data.component.js:98]: Removed.

File: src/views/record-browser/table-data/table-data.html (hunk 21 above)
Comment (ThomastheBicyclist) [inline src/views/record-browser/table-data/table-data.html:12]: Done.
